controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,171 @@
+//+kubebuilder:rbac:groups=metal3.io,resources=firmwareschemas/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *HostFirmwareSettingsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+
+	reqLogger := r.Log.WithValues("hostfirmwaresettings", req.NamespacedName)
+	reqLogger.Info("start")
+
+	// Fetch the HostFirmwareSettings
+	hfs := &metal3v1alpha1.HostFirmwareSettings{}
+	if err = r.Get(ctx, req.NamespacedName, hfs); err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after
+			// reconcile request. Return and don't requeue
+			reqLogger.Info("could not find host firmware settings")
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrap(err, "could not load host firmware settings")
+	}
+
+	// Check if settings has been cleared during pivot, and if so, restore firmwareSchema from annotation
+	if hfs.Status.FirmwareSchema == nil {
+		if _, ok := hfs.Annotations[metal3v1alpha1.FirmwareSchemaNameAnnotation]; ok {
+			hfs.Status.FirmwareSchema = &metal3v1alpha1.SchemaReference{
+				Name:      hfs.Annotations[metal3v1alpha1.FirmwareSchemaNameAnnotation],
+				Namespace: hfs.Annotations[metal3v1alpha1.FirmwareSchemaNamespaceAnnotation]}
+
+			if r.Status().Update(context.TODO(), hfs) != nil {
+				reqLogger.Info("could not update hostfirmwaresettings", "hostfirmwaresettings", req.NamespacedName)
+			}
+		}
+	}
+
+	// The Settings field cleared during pivot will be updated by baremetalhost controller which accesses Ironic via provisioner
+	if hfs.Status.Settings == nil {
+		reqLogger.Info("hostfirmwaresettings settings field has been cleared", "hostfirmwaresettings", req.NamespacedName)
+		return ctrl.Result{}, nil
+	}
+
+	// Get the referenced schema
+	var schema *metal3v1alpha1.FirmwareSchema
+	if hfs.Status.FirmwareSchema != nil {
+		schema = &metal3v1alpha1.FirmwareSchema{}
+		if err = r.Get(ctx, client.ObjectKey{
+			Namespace: hfs.Status.FirmwareSchema.Namespace,
+			Name:      hfs.Status.FirmwareSchema.Name}, schema); err != nil {
+			if k8serrors.IsNotFound(err) {
+				reqLogger.Info("could not find firmware schema for host", "hostfirmwaresettings", req.NamespacedName)
+				return ctrl.Result{}, nil
+			}
+
+			// Error reading the object - requeue the request.
+			return ctrl.Result{}, errors.Wrap(err, "could not load firmware schema")
+		}
+	} else {
+		schema = nil
+	}
+
+	if err = ValidateHostFirmwareSettings(hfs, schema); err != nil {
+		reqLogger.Info("hostFirmwareSetting validation failed", "error", err.Error())
```

Comment:
Also this one is somehow confusing, especially considering that no action or status is reported to the user